pkg/piecestore/rpc/client/readerwriter.go

```diff
@@ -44,20 +44,25 @@ type StreamReader struct {
 
 // Read -- Read method for piece download stream
 func (s *StreamReader) Read(b []byte) (int, error) {
+
+	// Use overflow data if we have it
 	if len(s.overflowData) > 0 {
-		n := copy(b, s.overflowData)
-		s.overflowData = s.overflowData[:len(s.overflowData)-n]
+		n := copy(b, s.overflowData)        // Copy from overflow into buffer
+		s.overflowData = s.overflowData[n:] // Overflow is set to whatever remains
 		return n, nil
 	}
 
-	pieceData, err := s.stream.Recv()
+	// Receive data from server stream
+	msg, err := s.stream.Recv()
 	if err != nil {
 		return 0, err
 	}
 
-	n := copy(b, pieceData.Content)
+	// Copy data into buffer
+	n := copy(b, msg.Content)
 
-	if n < len(pieceData.Content) {
+	// If left over data save it into overflow variable for next read
+	if n < len(msg.Content) {
```

Comment:
probably slightly more robust to do s.overflowData = b[n:] in this case, but don't worry about changing it. inside this if branch len(b) and n are interchangable.